ciao-controller/cnci.go

```diff
@@ -0,0 +1,579 @@
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/01org/ciao/ciao-controller/types"
+	"github.com/01org/ciao/payloads"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+// CNCIState represents specific allowed state strings
+type CNCIState string
+
+var (
+	exited CNCIState = payloads.Exited
+	active CNCIState = payloads.Running
+	failed CNCIState = payloads.ExitFailed
+)
+
+type event string
+
+var (
+	added        event = "concentrator added"
+	startFailure event = "cnci start failure"
+	removed      event = "concentrator removed"
+)
+
+var cnciEventTimeout = (2 * time.Minute)
+
+// CNCI represents a cnci instance that manages a single subnet.
+type CNCI struct {
+	instance    *types.Instance
+	ctrl        *controller
+	eventCh     *chan event
+	eventChLock sync.Mutex
+	subnet      int
+	timer       *time.Timer
+}
+
+// CNCIManager is a structure which defines a manager for CNCI instances
+// TBD: managing multiple CNCI instances.
+type CNCIManager struct {
+	tenant string
+	ctrl   *controller
+
+	// there's no reason to have separate lock for each map.
+	cnciLock *sync.RWMutex
+
+	// this is a map of CNCI instance IDs to CNCI structs
+	cncis map[string]*CNCI
+
+	// this is a map of subnet (integer) to CNCI structs
+	subnets map[int]*CNCI
+}
+
+func (c *CNCI) stop() error {
+	err := transitionInstanceState(c.instance, payloads.Stopping)
+	if err != nil {
+		return err
+	}
+
+	err = c.ctrl.deleteInstance(c.instance.ID)
+	if err != nil {
+		return errors.Wrapf(err, "error deleting CNCI instance")
+	}
+
+	return nil
+}
+
+func waitForEventTimeout(ch chan event, e event, timeout time.Duration) error {
+	select {
+	case recv := <-ch:
+		if recv != e {
+			return fmt.Errorf("expecting %v got %v", e, recv)
+		}
+		return nil
+	case <-time.After(timeout):
+		return fmt.Errorf("timeout waiting for event %v", e)
+	}
+}
+
+func (c *CNCI) transitionState(to CNCIState) {
+	transitionInstanceState(c.instance, (string(to)))
+
+	// some state changes cause events
+	c.eventChLock.Lock()
+	ch := c.eventCh
```

Comment:
Is there a reason to use a pointer for eventCh?  nil is a valid value for channels.

The locks are needed I guess as a state transition might not be always be initiated by one of the methods in CNCIManager.  Is this correct?

Reply:
no, I thought I needed them because multiple go routines within cnci.go might modify the eventCh variable value. For example, a new channel is created when you are waiting to remove a subnet, as well as when you are waiting to add one. Can I remove the locks?

Reply:
maybe its sufficient to just use the cnciLock lock.

Reply:
No I think you're correct.  You need the locks as you're creating and deleting channels.